// docusaurus-theme-openapi-docs/lib/theme/ApiExplorer/Body/FormBodyItem/index.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  (function () {
    var ownKeys = function (o) {
      ownKeys =
        Object.getOwnPropertyNames ||
        function (o) {
          var ar = [];
          for (var k in o)
            if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
          return ar;
        };
      return ownKeys(o);
    };
    return function (mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null)
        for (var k = ownKeys(mod), i = 0; i < k.length; i++)
          if (k[i] !== "default") __createBinding(result, mod, k[i]);
      __setModuleDefault(result, mod);
      return result;
    };
  })();
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod };
  };
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = FormBodyItem;
const react_1 = __importStar(require("react"));
const FormFileUpload_1 = __importDefault(
  require("@theme/ApiExplorer/FormFileUpload")
);
const FormLabel_1 = __importDefault(require("@theme/ApiExplorer/FormLabel"));
const FormSelect_1 = __importDefault(require("@theme/ApiExplorer/FormSelect"));
const FormTextInput_1 = __importDefault(
  require("@theme/ApiExplorer/FormTextInput")
);
const LiveEditor_1 = __importDefault(require("@theme/ApiExplorer/LiveEditor"));
const hooks_1 = require("@theme/ApiItem/hooks");
const FileArrayFormBodyItem_1 = __importDefault(
  require("../FileArrayFormBodyItem")
);
const slice_1 = require("../slice");
const slice_2 = require("../../EncodingSelection/slice");
function FormBodyItem({
  schemaObject,
  id,
  schema,
  label,
  required,
  exampleValue,
  fieldEncoding,
}) {
  const dispatch = (0, hooks_1.useTypedDispatch)();
  // Parse comma-separated encoding contentType into selectable options
  const encodingOptions = fieldEncoding
    ? fieldEncoding
        .split(",")
        .map((t) => t.trim())
        .filter(Boolean)
    : [];
  const hasMultipleEncodings = encodingOptions.length > 1;
  // Initialize with the first declared content type
  const [selectedEncoding, setSelectedEncoding] = (0, react_1.useState)(
    encodingOptions[0] ?? ""
  );
  // Seed Redux with the first declared encoding on mount so the code snippet
  // reflects a content type immediately, even before the user interacts.
  // The empty dep array is intentional: `fieldEncoding` comes from a static
  // spec value that never changes for the lifetime of this component instance,
  // and re-seeding on every render would fight user selections.
  (0, react_1.useEffect)(() => {
    if (encodingOptions[0]) {
      dispatch(
        (0, slice_2.setFieldEncoding)({
          field: id,
          contentType: encodingOptions[0],
        })
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  const [value, setValue] = (0, react_1.useState)(() => {
    let initialValue = exampleValue ?? "";
    if (schemaObject.type === "object" && exampleValue) {
      initialValue = JSON.stringify(exampleValue, null, 2);
    }
    return initialValue;
  });
  (0, react_1.useEffect)(() => {
    if (value) {
      dispatch((0, slice_1.setStringFormBody)({ key: id, value }));
    } else {
      dispatch((0, slice_1.clearFormBodyKey)(id));
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  if (
    schemaObject.type === "array" &&
    schemaObject.items?.format === "binary"
  ) {
    return react_1.default.createElement(
      react_1.default.Fragment,
      null,
      label &&
        react_1.default.createElement(FormLabel_1.default, {
          label: label,
          required: required,
        }),
      react_1.default.createElement(FileArrayFormBodyItem_1.default, {
        id: id,
        description: schemaObject.description,
      })
    );
  }
  if (schemaObject.format === "binary") {
    return react_1.default.createElement(
      react_1.default.Fragment,
      null,
      label &&
        react_1.default.createElement(FormLabel_1.default, {
          label: label,
          required: required,
        }),
      hasMultipleEncodings &&
        react_1.default.createElement(
          "div",
          { style: { marginTop: "0.5rem" } },
          react_1.default.createElement(FormSelect_1.default, {
            label: "Content-Type",
            options: encodingOptions,
            value: selectedEncoding,
            onChange: (e) => {
              const ct = e.target.value;
              setSelectedEncoding(ct);
              dispatch(
                (0, slice_2.setFieldEncoding)({ field: id, contentType: ct })
              );
            },
          })
        ),
      react_1.default.createElement(FormFileUpload_1.default, {
        placeholder: schemaObject.description || id,
        onChange: (file) => {
          if (file === undefined) {
            dispatch((0, slice_1.clearFormBodyKey)(id));
            return;
          }
          dispatch(
            (0, slice_1.setFileFormBody)({
              key: id,
              value: {
                src: `/path/to/${file.name}`,
                content: file,
              },
            })
          );
        },
      })
    );
  }
  if (schemaObject.type === "object") {
    return react_1.default.createElement(
      react_1.default.Fragment,
      null,
      label &&
        react_1.default.createElement(FormLabel_1.default, {
          label: label,
          required: required,
        }),
      react_1.default.createElement(
        LiveEditor_1.default,
        {
          action: (code) =>
            dispatch((0, slice_1.setStringFormBody)({ key: id, value: code })),
        },
        value
      )
    );
  }
  if (
    schemaObject.enum &&
    schemaObject.enum.every((value) => typeof value === "string")
  ) {
    return react_1.default.createElement(FormSelect_1.default, {
      label: label,
      required: required,
      value: value,
      options: ["---", ...schemaObject.enum],
      onChange: (e) => {
        const val = e.target.value;
        setValue(val);
        if (val === "---") {
          dispatch((0, slice_1.clearFormBodyKey)(id));
        } else {
          dispatch(
            (0, slice_1.setStringFormBody)({
              key: id,
              value: val,
            })
          );
        }
      },
    });
  }
  // TODO: support all the other types.
  return react_1.default.createElement(FormTextInput_1.default, {
    label: label,
    required: required,
    value: value,
    paramName: id,
    isRequired: Array.isArray(schema.required) && schema.required.includes(id),
    placeholder: schemaObject.description || id,
    onChange: (e) => {
      setValue(e.target.value);
      dispatch(
        (0, slice_1.setStringFormBody)({ key: id, value: e.target.value })
      );
    },
  });
}
