// docusaurus-theme-openapi-docs/lib/theme/ApiExplorer/FormTextInput/index.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  (function () {
    var ownKeys = function (o) {
      ownKeys =
        Object.getOwnPropertyNames ||
        function (o) {
          var ar = [];
          for (var k in o)
            if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
          return ar;
        };
      return ownKeys(o);
    };
    return function (mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null)
        for (var k = ownKeys(mod), i = 0; i < k.length; i++)
          if (k[i] !== "default") __createBinding(result, mod, k[i]);
      __setModuleDefault(result, mod);
      return result;
    };
  })();
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod };
  };
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
const react_1 = __importStar(require("react"));
const Translate_1 = require("@docusaurus/Translate");
const error_message_1 = require("@hookform/error-message");
const FormLabel_1 = __importDefault(require("@theme/ApiExplorer/FormLabel"));
const translationIds_1 = require("@theme/translationIds");
const clsx_1 = __importDefault(require("clsx"));
const react_hook_form_1 = require("react-hook-form");
function FormTextInput({
  isRequired,
  value,
  placeholder,
  password,
  onChange,
  paramName,
  label,
  type,
  required,
}) {
  const id = (0, react_1.useId)();
  placeholder = placeholder?.split("\n")[0];
  const {
    register,
    formState: { errors },
  } = (0, react_hook_form_1.useFormContext)();
  const showErrorMessage = errors?.[paramName]?.message;
  return react_1.default.createElement(
    react_1.default.Fragment,
    null,
    label &&
      react_1.default.createElement(FormLabel_1.default, {
        htmlFor: id,
        label: label,
        type: type,
        required: required,
      }),
    paramName
      ? react_1.default.createElement("input", {
          ...register(paramName, {
            required: isRequired
              ? (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_FORM.FIELD_REQUIRED,
                  message: "This field is required",
                })
              : false,
          }),
          id: label ? id : undefined,
          className: (0, clsx_1.default)("openapi-explorer__form-item-input", {
            error: showErrorMessage,
          }),
          type: password ? "password" : "text",
          placeholder: placeholder,
          title: placeholder,
          value: value,
          onChange: onChange,
          autoComplete: "off",
        })
      : react_1.default.createElement("input", {
          id: label ? id : undefined,
          className: "openapi-explorer__form-item-input",
          type: password ? "password" : "text",
          placeholder: placeholder,
          title: placeholder,
          value: value,
          onChange: onChange,
          autoComplete: "off",
        }),
    showErrorMessage &&
      react_1.default.createElement(error_message_1.ErrorMessage, {
        errors: errors,
        name: paramName,
        render: ({ message }) =>
          react_1.default.createElement(
            "div",
            { className: "openapi-explorer__input-error" },
            message
          ),
      })
  );
}
exports.default = FormTextInput;
