// docusaurus-plugin-openapi-docs/lib/openapi/createSchemaExample.test.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
Object.defineProperty(exports, "__esModule", { value: true });
const createSchemaExample_1 = require("./createSchemaExample");
describe("sampleFromSchema", () => {
    describe("const support", () => {
        it("should return default string value when const is not present", () => {
            const schema = {
                type: "string",
            };
            const context = { type: "request" };
            const result = (0, createSchemaExample_1.sampleFromSchema)(schema, context);
            expect(result).toBe("string");
        });
        it("should return const value when const is present", () => {
            const schema = {
                type: "string",
                const: "example",
            };
            const context = { type: "request" };
            const result = (0, createSchemaExample_1.sampleFromSchema)(schema, context);
            expect(result).toBe("example");
        });
        it("should handle anyOf with const values", () => {
            const schema = {
                type: "string",
                anyOf: [
                    {
                        type: "string",
                        const: "dog",
                    },
                    {
                        type: "string",
                        const: "cat",
                    },
                ],
            };
            const context = { type: "request" };
            const result = (0, createSchemaExample_1.sampleFromSchema)(schema, context);
            expect(result).toBe("dog");
        });
    });
    describe("allOf with incompatible types", () => {
        it("should return undefined when allOf contains incompatible types", () => {
            const schema = {
                allOf: [{ type: "string" }, { type: "integer" }],
            };
            const context = { type: "request" };
            const result = (0, createSchemaExample_1.sampleFromSchema)(schema, context);
            expect(result).toBeUndefined();
        });
        it("should handle incompatible allOf types in a property", () => {
            const schema = {
                type: "object",
                properties: {
                    numero: {
                        allOf: [{ type: "string" }, { type: "integer" }],
                    },
                    name: {
                        type: "string",
                    },
                },
            };
            const context = { type: "request" };
            const result = (0, createSchemaExample_1.sampleFromSchema)(schema, context);
            expect(result.name).toBe("string");
        });
    });
});
