// docusaurus-theme-openapi-docs/lib/theme/ApiExplorer/FormMultiSelect/index.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  (function () {
    var ownKeys = function (o) {
      ownKeys =
        Object.getOwnPropertyNames ||
        function (o) {
          var ar = [];
          for (var k in o)
            if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
          return ar;
        };
      return ownKeys(o);
    };
    return function (mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null)
        for (var k = ownKeys(mod), i = 0; i < k.length; i++)
          if (k[i] !== "default") __createBinding(result, mod, k[i]);
      __setModuleDefault(result, mod);
      return result;
    };
  })();
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod };
  };
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const FormLabel_1 = __importDefault(require("@theme/ApiExplorer/FormLabel"));
const clsx_1 = __importDefault(require("clsx"));
function FormMultiSelect({
  value,
  options,
  onChange,
  showErrors,
  label,
  type,
  required,
}) {
  const id = (0, react_1.useId)();
  if (options.length === 0) {
    return null;
  }
  let height;
  if (options.length < 6) {
    const selectPadding = 12 * 2;
    const rawHeight = options.length * 29;
    const innerMargins = 4 * options.length - 1;
    const outerMargins = 4 * 2;
    const mysteryScroll = 1;
    height =
      rawHeight + innerMargins + outerMargins + selectPadding + mysteryScroll;
  }
  return react_1.default.createElement(
    react_1.default.Fragment,
    null,
    label &&
      react_1.default.createElement(FormLabel_1.default, {
        htmlFor: id,
        label: label,
        type: type,
        required: required,
      }),
    react_1.default.createElement(
      "select",
      {
        id: label ? id : undefined,
        style: { height: height },
        className: (0, clsx_1.default)("openapi-explorer__multi-select-input", {
          error: showErrors,
        }),
        value: value,
        onChange: onChange,
        size: Math.min(6, options.length + 1),
        multiple: true,
      },
      options.map((option) => {
        return react_1.default.createElement(
          "option",
          { key: option, value: option },
          option
        );
      })
    )
  );
}
exports.default = FormMultiSelect;
