// docusaurus-theme-openapi-docs/lib/theme/RequestSchema/index.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod };
  };
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const BrowserOnly_1 = __importDefault(require("@docusaurus/BrowserOnly"));
const Translate_1 = require("@docusaurus/Translate");
const Details_1 = __importDefault(require("@theme/Details"));
const Markdown_1 = __importDefault(require("@theme/Markdown"));
const MimeTabs_1 = __importDefault(require("@theme/MimeTabs")); // Assume these components exist
const ResponseExamples_1 = require("@theme/ResponseExamples");
const Schema_1 = __importDefault(require("@theme/Schema"));
const SchemaTabs_1 = __importDefault(require("@theme/SchemaTabs"));
const SkeletonLoader_1 = __importDefault(require("@theme/SkeletonLoader"));
const TabItem_1 = __importDefault(require("@theme/TabItem"));
const translationIds_1 = require("@theme/translationIds");
const RequestSchemaComponent = ({ title, body, style }) => {
  if (
    body === undefined ||
    body.content === undefined ||
    Object.keys(body).length === 0 ||
    Object.keys(body.content).length === 0
  ) {
    return null;
  }
  const mimeTypes = Object.keys(body.content);
  if (mimeTypes.length > 1) {
    return react_1.default.createElement(
      MimeTabs_1.default,
      { className: "openapi-tabs__mime", schemaType: "request", lazy: true },
      mimeTypes.map((mimeType) => {
        const mediaTypeObject = body.content[mimeType];
        const firstBody = mediaTypeObject?.schema;
        const requestExamples = mediaTypeObject?.examples;
        const requestExample = mediaTypeObject?.example;
        if (
          firstBody === undefined ||
          (firstBody.properties &&
            Object.keys(firstBody.properties).length === 0)
        ) {
          return null;
        }
        return (
          // @ts-ignore
          react_1.default.createElement(
            TabItem_1.default,
            { key: mimeType, label: mimeType, value: mimeType },
            react_1.default.createElement(
              SchemaTabs_1.default,
              { className: "openapi-tabs__schema" },
              react_1.default.createElement(
                TabItem_1.default,
                { key: title, label: title, value: title },
                react_1.default.createElement(
                  "div",
                  { style: { marginTop: "1rem" } },
                  react_1.default.createElement(
                    Details_1.default,
                    {
                      className: "openapi-markdown__details mime",
                      "data-collapsed": false,
                      open: true,
                      style: style,
                      summary: react_1.default.createElement(
                        react_1.default.Fragment,
                        null,
                        react_1.default.createElement(
                          "summary",
                          null,
                          react_1.default.createElement(
                            "h3",
                            {
                              className:
                                "openapi-markdown__details-summary-header-body",
                            },
                            (0, Translate_1.translate)({
                              id: translationIds_1.OPENAPI_REQUEST.BODY_TITLE,
                              message: title,
                            }),
                            body.required === true &&
                              react_1.default.createElement(
                                "span",
                                { className: "openapi-schema__required" },
                                (0, Translate_1.translate)({
                                  id: translationIds_1.OPENAPI_SCHEMA_ITEM
                                    .REQUIRED,
                                  message: "required",
                                })
                              )
                          )
                        )
                      ),
                    },
                    react_1.default.createElement(
                      "div",
                      { style: { textAlign: "left", marginLeft: "1rem" } },
                      body.description &&
                        react_1.default.createElement(
                          "div",
                          {
                            style: { marginTop: "1rem", marginBottom: "1rem" },
                          },
                          react_1.default.createElement(
                            Markdown_1.default,
                            null,
                            body.description
                          )
                        )
                    ),
                    react_1.default.createElement(
                      "ul",
                      { style: { marginLeft: "1rem" } },
                      react_1.default.createElement(Schema_1.default, {
                        schema: firstBody,
                        schemaType: "request",
                        schemaPath: "requestBody",
                      })
                    )
                  )
                )
              ),
              firstBody &&
                !requestExample &&
                !requestExamples &&
                (0, ResponseExamples_1.ExampleFromSchema)({
                  schema: firstBody,
                  mimeType: mimeType,
                }),
              requestExamples &&
                (0, ResponseExamples_1.ResponseExamples)({
                  responseExamples: requestExamples,
                  mimeType,
                }),
              requestExample &&
                (0, ResponseExamples_1.ResponseExample)({
                  responseExample: requestExample,
                  mimeType,
                })
            )
          )
        );
      })
    );
  }
  const randomFirstKey = mimeTypes[0];
  const mediaTypeObject = body.content[randomFirstKey];
  const firstBody = mediaTypeObject?.schema ?? body.content[randomFirstKey];
  const requestExamples = mediaTypeObject?.examples;
  const requestExample = mediaTypeObject?.example;
  if (firstBody === undefined) {
    return null;
  }
  return react_1.default.createElement(
    MimeTabs_1.default,
    { className: "openapi-tabs__mime", schemaType: "request" },
    react_1.default.createElement(
      TabItem_1.default,
      { label: randomFirstKey, value: `${randomFirstKey}-schema` },
      react_1.default.createElement(
        SchemaTabs_1.default,
        { className: "openapi-tabs__schema" },
        react_1.default.createElement(
          TabItem_1.default,
          { key: title, label: title, value: title },
          react_1.default.createElement(
            Details_1.default,
            {
              className: "openapi-markdown__details mime",
              "data-collapsed": false,
              open: true,
              style: style,
              summary: react_1.default.createElement(
                react_1.default.Fragment,
                null,
                react_1.default.createElement(
                  "summary",
                  null,
                  react_1.default.createElement(
                    "h3",
                    {
                      className:
                        "openapi-markdown__details-summary-header-body",
                    },
                    (0, Translate_1.translate)({
                      id: translationIds_1.OPENAPI_REQUEST.BODY_TITLE,
                      message: title,
                    }),
                    firstBody.type === "array" &&
                      react_1.default.createElement(
                        "span",
                        { style: { opacity: "0.6" } },
                        " array"
                      ),
                    body.required &&
                      react_1.default.createElement(
                        "strong",
                        { className: "openapi-schema__required" },
                        (0, Translate_1.translate)({
                          id: translationIds_1.OPENAPI_SCHEMA_ITEM.REQUIRED,
                          message: "required",
                        })
                      )
                  )
                )
              ),
            },
            react_1.default.createElement(
              "div",
              { style: { textAlign: "left", marginLeft: "1rem" } },
              body.description &&
                react_1.default.createElement(
                  "div",
                  { style: { marginTop: "1rem", marginBottom: "1rem" } },
                  react_1.default.createElement(
                    Markdown_1.default,
                    null,
                    body.description
                  )
                )
            ),
            react_1.default.createElement(
              "ul",
              { style: { marginLeft: "1rem" } },
              react_1.default.createElement(Schema_1.default, {
                schema: firstBody,
                schemaType: "request",
                schemaPath: "requestBody",
              })
            )
          )
        ),
        firstBody &&
          !requestExample &&
          !requestExamples &&
          (0, ResponseExamples_1.ExampleFromSchema)({
            schema: firstBody,
            mimeType: randomFirstKey,
          }),
        requestExamples &&
          (0, ResponseExamples_1.ResponseExamples)({
            responseExamples: requestExamples,
            mimeType: randomFirstKey,
          }),
        requestExample &&
          (0, ResponseExamples_1.ResponseExample)({
            responseExample: requestExample,
            mimeType: randomFirstKey,
          })
      )
    )
  );
};
const RequestSchema = (props) => {
  return react_1.default.createElement(
    BrowserOnly_1.default,
    {
      fallback: react_1.default.createElement(SkeletonLoader_1.default, {
        size: "sm",
      }),
    },
    () => {
      return react_1.default.createElement(RequestSchemaComponent, {
        ...props,
      });
    }
  );
};
exports.default = RequestSchema;
