// docusaurus-plugin-openapi-docs/lib/sidebars/index.test.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = __importDefault(require("./index"));
describe("generateSidebarSlice", () => {
    describe("tagGroup with overlapping tags", () => {
        const mockApiItems = [
            {
                type: "api",
                id: "get-books",
                unversionedId: "get-books",
                title: "Get Books",
                description: "",
                source: "",
                sourceDirName: "",
                permalink: "/get-books",
                frontMatter: {},
                api: {
                    method: "get",
                    path: "/books",
                    tags: ["Books", "Deprecated"],
                    jsonRequestBodyExample: "",
                    info: { title: "Test API", version: "1.0.0" },
                },
            },
        ];
        const mockTags = [
            [
                { name: "Books", description: "Book operations" },
                { name: "Deprecated", description: "Deprecated endpoints" },
            ],
        ];
        const mockTagGroups = [
            { name: "Library", tags: ["Books"] },
            { name: "Deprecation", tags: ["Deprecated"] },
        ];
        function collectKeys(obj) {
            const keys = [];
            JSON.stringify(obj, (k, v) => {
                if (k === "key" && typeof v === "string") {
                    keys.push(v);
                }
                return v;
            });
            return keys;
        }
        it("should generate unique keys for items appearing in multiple tagGroups", () => {
            const result = (0, index_1.default)({ groupPathsBy: "tagGroup" }, { outputDir: "docs/test", specPath: "" }, mockApiItems, mockTags, "", mockTagGroups);
            const keys = collectKeys(result);
            expect(keys.length).toBeGreaterThan(0);
            expect(new Set(keys).size).toBe(keys.length);
        });
        it("should include tagGroup name in keys to differentiate same items", () => {
            const result = (0, index_1.default)({ groupPathsBy: "tagGroup" }, { outputDir: "docs/test", specPath: "" }, mockApiItems, mockTags, "", mockTagGroups);
            const keys = collectKeys(result);
            expect(keys.filter((k) => k.includes("library")).length).toBeGreaterThan(0);
            expect(keys.filter((k) => k.includes("deprecation")).length).toBeGreaterThan(0);
        });
    });
});
